### PR TITLE
Add tutorial for implementing shopping cart

### PR DESCRIPTION
**Pull request by andreia-oca**

### CODE REVIEW

**Comment by andreia-oca** [comment]
A live preview is available at 
https://gray-just-nightingale.dev.app.genez.io/blog/how-to-implement-a-shopping-cart-feature-using-typescript-redis-and-react/

Feel free to leave comments and improvements in the PR.

File: content/posts/implement_shopping_cart.md [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+
+A better option is to use a Redis database on the server-side to store items persistently and retrieve them very fast when needed.
+
+This tutorial will show you how to harness the speed of a Redis database to build a shopping cart functionality using TypeScript/NodeJs.
+
+{{< details "Expand this section to get more information on what's Redis" >}}
+
+Redis is an open-source, in-memory database that store key, values pairs. Redis is a highly versatile and performant database system. One of Redis's key features is its speed, as it primarily operates in-memory, enabling rapid data access. It is widely used for caching, real-time analytics, message brokering, and other applications where low-latency and high-throughput data storage and retrieval are crucial. Redis also provides persistence options, allowing data to be saved to disk for durability. Its simplicity, efficiency, and support for advanced data structures make Redis a popular choice for applications requiring fast and scalable data storage and retrieval.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:47]:
```suggestion
Redis is an open-source, in-memory database that stores key, values pairs. Redis is a highly versatile and performant database system. One of Redis's key features is its speed, as it primarily operates in-memory, enabling rapid data access. It is widely used for caching, real-time analytics, message brokering, and other applications where low-latency and high-throughput data storage and retrieval are crucial. Redis also provides persistence options, allowing data to be saved to disk for durability. Its simplicity, efficiency, and support for advanced data structures make Redis a popular choice for applications requiring fast and scalable data storage and retrieval.
```

File: content/posts/implement_shopping_cart.md [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+  );
+```
+
+This state should be updated each time an item is added or removed from the cart. When the cart is cleared, the state should be reset to 0.
+
+### Add spinners animation for loading states
+
+To improve the UI even further, you can add spinners to indicate loading states on the buttons that are sending requests to the backend.
+
+Each button should have a its own loading state to indicate that the request is being processed.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:590]: A determiner cannot be combined with a possessive pronoun. Did you mean simply 'a' or 'its'?

File: content/posts/implement_shopping_cart.md [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+      )}
+    </Button>
+);
+```
+
+## Conclusion
+
+Congratulations! 🥳
+
+The codebase for this tutorial is open-source and you can find it on the [GitHub repository](https://github.com/genez-io/genezio-examples/tree/main/typescript/shopping-cart).
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:628]:
```suggestion
The codebase for this tutorial is open-source, and you can find it on the [GitHub repository](https://github.com/genez-io/genezio-examples/tree/main/typescript/shopping-cart).
```

File: content/posts/implement_shopping_cart.md [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+`nvm` will automatically install the corresponding `npm` version.
+
+## Getting Started
+
+Firstly, you'll need to create a new project with a server and a client.
+The server will be responsible to communicate (store, retrieve, update, delete) with the Redis database.
+The client will render the products available in the store and allow the user to add them to the shopping cart.
+
+To get started with a template, install `genezio` using `npm` and run it in your terminal.
+Later on, `genezio` come in handy to deploy and host your web applications in the cloud.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:90]: Do you mean `genezio will come in handy`?

File: content/posts/implement_shopping_cart.md [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+
+After installing `genezio`, you can scaffold the application by running the following command:
+
+```bash
+genezio
+```
+
+The command above will get you through a series of questions to help you customize and prepare your project for production deployment.
+
+Your terminal should similar to the following output:
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:104]:
```suggestion
Your terminal should look similar to the following output:
```

File: content/posts/implement_shopping_cart.md [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+
+No subdomain specified in the genezio.yaml configuration file. We will provide a random one for you.
+Frontend successfully deployed at https://amber-cooperative-blackbird.app.genez.io.
+```
+
+The output from the `genezio` command will contain a randomly-assigned subdomain where your project was deployed for testing purposes.
+
+## Setting up the backend
+
+Now that you have a template to start from, fire up your favorite code editor and open project's directory.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:134]:
```suggestion
Now that you have a template to start from, fire up your favorite code editor and open the project's directory.
```

File: content/posts/implement_shopping_cart.md [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+Let's create a Redis database and connect it to your web application using a `.env` file.
+
+Luckily, genezio makes it very easy to add an Upstash Redis integration to your project.
+Go to the [genezio dashboard](https://app.genez.io) and add an Upstash Redis integration to your project.
+
+![Alt text](/images/blog/shopping_cart_example/add_integration.png)
+
+If you get stuck at this step, you can follow the [Upstash Redis integration tutorial](https://docs.genez.io/genezio-documentation/integrations/upstash-redis).
+
+After adding the Upstash Redis integration, copy the environment variable from the `.env` section. Go back to your `server` directory, create a new file called `.env` and paste the contents there.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:214]: I think somebody might think that they should copy the *** star output and paste it into the `.env` file. You should mention that they should explicitly use the copy button to disclose their tokens.

File: content/posts/implement_shopping_cart.md [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+```
+
+You can test your application from the genezio dashboard by following the URL exposed in the terminal by `genezio local`.
+
+```
+~ genezio local
+Test your code at https://app.genez.io/test-interface/local?port=8083
+```
+
+From the genezio dashboard you can sent request to your backend classes and see the response.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:232]:
```suggestion
From the genezio dashboard you can send requests to your backend classes and see the responses.
```

File: content/posts/implement_shopping_cart.md [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+
+![Alt text](/images/blog/shopping_cart_example/test_backend.png)
+
+You'll also need 2 more methods in the backend source code - a method to remove only an item from the cart and a method to delete the entire cart.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:236]:
```suggestion
 You'll also need 2 more methods in the backend source code - a method to remove only one item from the cart and a method to delete the entire cart.
```

File: content/posts/implement_shopping_cart.md [hunk 10]
```diff
@@ -0,0 +1,630 @@
...
+    const updatedCartData = await ShoppingCartService.getCart(token);
+    setCartData(updatedCartData);
+  };
+```
+
+## Deploying the application
+
+Your application is now ready to be deployed to the cloud to be used by your clients.
+
+Before deploying your application, you can change the randomly-assigned subdomain to something more meaningly for your application.
```
Comment (costinsin) [inline content/posts/implement_shopping_cart.md:520]:
```suggestion
Before deploying your application, you can change the randomly-assigned subdomain to something more meaningful for your application.
```

**Review by costinsin — CHANGES_REQUESTED** [review]
I felt like the frontend part is a little bit rushed and assumes that the user knows what's a React application. Maybe you can leave a link to the code before the section starts and state that if they are ever lost they can anytime check the code or copy it directly.

**Pull request merged**